packages/plugin-core/src/commands/CreateSchemaFromHierarchyCommand.ts

```diff
@@ -0,0 +1,528 @@
+
+  depth() {
+    return this.tokens.length;
+  }
+
+  topId() {
+    return this.levels[0].topId();
+  }
+
+  /**
+   * Levels of the hierarchy that we deem as viable options for creating a schema for.
+   * We remove the first level since having something like `*.h1.h2` with `*` at the
+   * beginning will match all hierarchies. Therefore we slice off the first level.
+   *
+   * We also slice off the last level since we need child notes to exist to
+   * create a meaningful hierarchy.
+   * */
+  getSchemaebleLevels() {
+    return this.levels.slice(1, this.levels.length - 1);
+  }
+}
+
+export type SchemaCandidate = {
+  note: NoteProps;
+  label: string;
+  detail: string;
+};
+
+function isDescendentOf(
+  descendentCandidate: SchemaCandidate,
+  ancestorCandidate: SchemaCandidate
+) {
+  const isChild = descendentCandidate.note.fname.startsWith(
+    ancestorCandidate.note.fname + "."
+  );
+  return isChild;
+}
+
+function createCandidatesMapByFname(items: readonly SchemaCandidate[]) {
+  return new Map(items.map((item) => [item.note.fname, item]));
+}
+
+function getUriFromSchema(schema: SchemaModuleProps) {
+  const vaultPath = vault2Path({
+    vault: schema.vault,
+    wsRoot: getDWorkspace().wsRoot,
+  });
+  const uri = Uri.file(
+    SchemaUtils.getPath({ root: vaultPath, fname: schema.fname })
+  );
+  return uri;
+}
+
+function getSchemaUri(vault: DVault, schemaName: string) {
+  const vaultPath = vault2Path({ vault, wsRoot: getDWorkspace().wsRoot });
+  const uri = Uri.file(
+    SchemaUtils.getPath({ root: vaultPath, fname: schemaName })
+  );
+  return uri;
+}
+
+/**
+ * Encapsulates methods that are responsible for user interaction when
+ * asking user for input data.
+ * */
+export class UserQueries {
+  static async haveUserPickSchemaFileName(
+    hierarchyLevel: HierarchyLevel,
+    vault: DVault
+  ): Promise<string | undefined> {
+    let alreadyExists = false;
+    let schemaName: string | undefined;
+
+    do {
+      // eslint-disable-next-line no-await-in-loop
+      schemaName = await VSCodeUtils.showInputBox({
+        value: hierarchyLevel.getDefaultSchemaName(),
+      });
+
+      if (!schemaName) {
+        // Cancelled.
+        return schemaName;
+      }
+
+      alreadyExists = fs.existsSync(getSchemaUri(vault, schemaName).fsPath);
+      if (alreadyExists) {
+        vscode.window.showInformationMessage(
+          `Schema with name '${schemaName}' already exists. Pick a different name or cancel.`
+        );
+      }
+    } while (alreadyExists);
+
+    return schemaName;
+  }
+
+  static async haveUserSelectHierarchyLevel(currDocFsPath: string) {
+    const hierarchy = new Hierarchy(path.basename(currDocFsPath, ".md"));
+
+    if (hierarchy.depth() <= 2) {
+      // We require some depth to the hierarchy to be able to choose a variance
+      // pattern within in it. More info within Hierarchy object.
+      await vscode.window.showErrorMessage(
+        `Pick a note deeper within the note hierarchy.`
```

Comment:
Should we be explicit in the error message that the depth must be greater than or equal to 2? Instead of saying 'deeper', which may be ambiguous.